### PR TITLE
Hashicorp Vault AWS Authentication

### PR DESCRIPTION
**Pull request by nniehoff**
This PR Implements Vault AWS Authentication completing #34.  This PR is also based on the work I did in #35.  I refactored a couple of methods to clean up the code a bit here.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -162,10 +162,14 @@ PLUGINS_CONFIG = {
 ```
 
 - `url` - (required) The URL to the HashiCorp Vault instance (e.g. `http://localhost:8200`).
-- `auth_method` - (optional / defaults to "token") The method used to authenticate against the HashiCorp Vault instance. Either `"token"` or `"approle"`.
+- `auth_method` - (optional / defaults to "token") The method used to authenticate against the HashiCorp Vault instance. Either `"approle"`, `"aws"`, `"kubernetes"` or `"token"`.
+- `ca_cert` - (optional) Path to a PEM formatted CA certificatee to use when verifying the vault connection.  Set to `False` to ignore ssl verification (not recommended) or `True` to use the system certificates.
```
Comment (glennmatthews) [inline README.md:166]:
```suggestion
- `ca_cert` - (optional) Path to a PEM formatted CA certificate to use when verifying the Vault connection.  Can alternatively be set to `False` to ignore SSL verification (not recommended) or `True` to use the system certificates.
```

Does this have a default value?

File: nautobot_secrets_providers/providers/hashicorp.py [hunk 2]
```diff
@@ -41,19 +49,101 @@ class ParametersForm(BootstrapMixin, forms.Form):
...
+    def validate_vault_settings(cls, secret=None):
+        """Validate the vault settings."""
         # This is only required for HashiCorp Vault therefore not defined in
         # `required_settings` for the plugin config.
         plugin_settings = settings.PLUGINS_CONFIG["nautobot_secrets_providers"]
         if "hashicorp_vault" not in plugin_settings:
             raise exceptions.SecretProviderError(secret, cls, "HashiCorp Vault is not configured!")
 
-        vault_settings = plugin_settings["hashicorp_vault"]
+        vault_settings = plugin_settings.get("hashicorp_vault")
```
Comment (glennmatthews) [inline nautobot_secrets_providers/providers/hashicorp.py:60]:
```suggestion
        vault_settings = plugin_settings.get("hashicorp_vault", {})
```

File: nautobot_secrets_providers/providers/hashicorp.py [hunk 3]
```diff
@@ -41,19 +49,101 @@ class ParametersForm(BootstrapMixin, forms.Form):
...
         if "url" not in vault_settings:
             raise exceptions.SecretProviderError(secret, cls, "HashiCorp Vault configuration is missing a url")
 
+        if auth_method not in AUTH_METHOD_CHOICES:
+            raise exceptions.SecretProviderError(secret, cls, f"HashiCorp Vault Auth Method {auth_method} is invalid!")
+
+        if auth_method == "aws":
+            if not boto3:
+                raise exceptions.SecretProviderError(
+                    secret, cls, "HashiCorp Vault AWS Auth Method requires the boto3 plugin!"
```
Comment (glennmatthews) [inline nautobot_secrets_providers/providers/hashicorp.py:72]:
```suggestion
                    secret, cls, "HashiCorp Vault AWS Authentication Method requires the boto3 library!"
```

File: README.md [hunk 4]
```diff
@@ -162,10 +162,14 @@ PLUGINS_CONFIG = {
...
+- `k8s_token_path` - (optional) Path to the kubernetes service account token file.  Defaults to "/var/run/secrets/kubernetes.io/serviceaccount/token".
 - `token` - (optional) Required when `"auth_method": "token"` or `auth_method` is not supplied. The token for authenticating the client with the HashiCorp Vault instance. As with other sensitive service credentials, we recommend that you provide the token value as an environment variable and retrieve it with `{"token": os.getenv("NAUTOBOT_HASHICORP_VAULT_TOKEN")}` rather than hard-coding it in your `nautobot_config.py`.
+- `role_name` - (optional) Required when `"auth_method": "kubernetes"`.  The Vault Kubernetes role to assume which the pod's service account has access to.
```
Comment (glennmatthews) [inline README.md:169]: Would `k8s_role_name` be at all preferable to make this less ambiguous with `role_id`?

(Also, I see in the code below that we use `role_name` also for `aws` auth_method -- does the above need to be updated?

File: nautobot_secrets_providers/tests/test_providers.py [hunk 5]
```diff
@@ -202,3 +248,143 @@ def test_retrieve_invalid_key(self, requests_mocker):
...
+                    "auth_method": "kubernetes",
+                    "k8s_token_path": k8s_token_path,
+                },
+            },
+        }
+
+        # Test without specifying a role_name
+        with self.settings(PLUGINS_CONFIG=new_plugins_config):
+            with self.assertRaises(exceptions.SecretProviderError):
+                self.provider.get_client(self.secret)
```
Comment (glennmatthews) [inline nautobot_secrets_providers/tests/test_providers.py:271]: Might be good to catch the exception and assert that it has the correct/expected text in each such case?

```suggestion
            with self.assertRaises(exceptions.SecretProviderError) as context:
                self.provider.get_client(self.secret)
            self.assertEqual(str(context.exception), "...")
```

File: README.md (hunk 1 above)
Comment (nniehoff) [reply README.md:166]: no, default is None

File: README.md (hunk 4 above)
Comment (nniehoff) [reply README.md:169]: Within Vault there are role names for k8s and for was.  There are role ids and secret ids only for the auth method of "AppRole".  I do not know why AWS authentication IAM login does not require `role_name` but according to the [docs](https://hvac.readthedocs.io/en/stable/source/hvac_api_auth_methods.html#hvac.api.auth_methods.Aws.iam_login) it does not.  I prefer to leave `role_id` and `role_name` as they are as they have contextual meaning with in Hashicorp Vault.

Comment (glennmatthews) [reply README.md:169]: So perhaps:

```suggestion
- `role_name` - (optional) Required when `"auth_method": "kubernetes"`, optional when `"auth_method": "aws"`.  The Vault Kubernetes role or AWS role to assume which the pod's service account has access to.
```

File: nautobot_secrets_providers/providers/hashicorp.py [hunk 6]
```diff
@@ -41,19 +49,101 @@ class ParametersForm(BootstrapMixin, forms.Form):
...
+                    with open(k8s_token_path, "r", encoding="utf-8") as token_file:
+                        jwt = token_file.read()
+                    client.auth.kubernetes.login(role=vault_settings["role_name"], jwt=jwt, **login_kwargs)
+                elif auth_method == "aws":
+                    session = boto3.Session()
+                    aws_creds = session.get_credentials()
+                    client.auth.aws.iam_login(
+                        access_key=aws_creds.access_key,
+                        secret_key=aws_creds.secret_key,
+                        session_token=aws_creds.token,
```
Comment (glennmatthews) [inline nautobot_secrets_providers/providers/hashicorp.py:128]: Might be good to explicitly document in README that the AWS credentials *cannot* be specified via `vault_settings`, they must be pre-configured for AWS directly (and link to the appropriate docs)?

Alternately, should we for consistency with other methods at least provide an *option* to set the AWS credentials via `vault_settings`?

File: nautobot_secrets_providers/tests/test_providers.py [hunk 7]
```diff
@@ -199,6 +253,198 @@ def test_retrieve_invalid_key(self, requests_mocker):
...
+                'SecretProviderError: Secret "hello-hashicorp" (provider "HashiCorpVaultSecretsProvider"): HashiCorp Vault Access Denied (auth_method: kubernetes). Error: , on post http://localhost:8200/v1/auth/kubernetes/login',
+            )
+
+            # Test Invalid Request
+            requests_mocker.register_uri(method="POST", url=f"{vault_url}/v1/auth/kubernetes/login", status_code=400)
+            with self.assertRaises(exceptions.SecretProviderError) as err:
+                self.provider.get_client(self.secret)
+            self.assertEqual(
+                str(err.exception),
+                'SecretProviderError: Secret "hello-hashicorp" (provider "HashiCorpVaultSecretsProvider"): HashiCorp Vault Login failed (auth_method: kubernetes). Error: , on post http://localhost:8200/v1/auth/kubernetes/login',
```
Comment (glennmatthews) [inline nautobot_secrets_providers/tests/test_providers.py:317]: Is it expected that in actual practice we will have a blank `Error: ,` text in the exception? Or is the request mock just not providing an accurate representation of the error we'd get in practice?

File: nautobot_secrets_providers/tests/test_providers.py [hunk 8]
```diff
@@ -199,6 +253,198 @@ def test_retrieve_invalid_key(self, requests_mocker):
...
+                'SecretProviderError: Secret "hello-hashicorp" (provider "HashiCorpVaultSecretsProvider"): HashiCorp Vault Access Denied (auth_method: aws). Error: , on post http://localhost:8200/v1/auth/aws/login',
+            )
+
+            # Test Invalid Request
+            requests_mocker.register_uri(method="POST", url=f"{vault_url}/v1/auth/aws/login", status_code=400)
+            with self.assertRaises(exceptions.SecretProviderError) as err:
+                self.provider.get_client(self.secret)
+            self.assertEqual(
+                str(err.exception),
+                'SecretProviderError: Secret "hello-hashicorp" (provider "HashiCorpVaultSecretsProvider"): HashiCorp Vault Login failed (auth_method: aws). Error: , on post http://localhost:8200/v1/auth/aws/login',
```
Comment (glennmatthews) [inline nautobot_secrets_providers/tests/test_providers.py:449]: Same question here regarding `Error: ,` in the test.

File: README.md (hunk 4 above)
Comment (nniehoff) [reply README.md]: Changed to `The Vault Kubernetes role or Vault AWS role` as I don't want these to be confused with either Kubernetes cluster roles or AWS IAM roles.  Term overloaded successfully

File: nautobot_secrets_providers/tests/test_providers.py (hunk 7 above)
Comment (nniehoff) [reply nautobot_secrets_providers/tests/test_providers.py:317]: The mock isn't providing a correct error response

File: nautobot_secrets_providers/providers/hashicorp.py (hunk 6 above)
Comment (nniehoff) [reply nautobot_secrets_providers/providers/hashicorp.py:128]: from a security perspective, and from and AWS best practices stance, I'll just say not just no but absolutely not!  We handle the AWS authentication the same way as AWS secrets manager, just to connect to Vault instead.  I'll update the docs.

**Review by glennmatthews — APPROVED** [review]
Nice work!

**Review by bryanculver — APPROVED** [review]

**Pull request merged**